모달이나 채널리스트의 상태는 지역에서 관리로 수정, 왼쪽 사이드 바에 dm을 추가, 쓰레드 리스트 헤더 구현

hansanguk0222 opened the pull request:
# 오늘 한 일
### 1. 모달을 나타내는 상태는 지역에서 관리하도록 수정
- useState를 모달에 내리는 방법으록 구현

### 2. 왼쪽 사이드 바에 dm을 추가
- channelList를 이용하여서 channelType을 주고 채널인지 dm인지 분리함

### 3. 쓰레드 리스트 헤더 구현
- 채널인지 dm인지 구분하여서 필요한 요소를 따로 추가하도록 설계

### 결과
![Peek 2020-12-01 23-22](https://user-images.githubusercontent.com/50985451/100752643-2e01b000-342c-11eb-8144-7c750569461c.gif)

#### 오늘은 리펙토링만 하다가 끝이 난 거 같네요. 내일 더욱 달려보겠습니다. 우선 일단 작업한 것 날립니다. dm은 채널 이름은 임시로 넣었습니다!

hansanguk0222 committed:
[refactor] 필요 없는 부분 삭제 및 라우터 기반 동작 구현(라우터 기반 동작은 미완성)

hansanguk0222 committed:
[feat] 오른쪽 디테일 바를 공통 요소로 분리 (RightSideBar의 반환 타입을 아직 정하지 못함)

hansanguk0222 committed:
[refactor] 모달을 키거나 채널을 키는 상태는 지역으로 관리하기 위하여 해당 부분을 삭제

hansanguk0222 committed:
[refactor] 모달을 부르는 버튼에서 useState 함수를 받아서 모달창을 끄고 닫게 동작하도록 함

hansanguk0222 committed:
[feat] 채널 타입에 따라서 DM인지 채널인지 왼쪽 사이드바에 뿌려주고 그에 따른 쓰레드 리스트 헤더도 다르게 보여줌

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/ChannelList.tsx:
```diff
@@ -7,28 +7,38 @@ import { useChannel, useAuth } from '@/hooks';
   useEffect(() => {
     callAPI();
-  }, [dispatch]);
+  }, []);
```
callAPI를 따로 선언하지 않고, dispatch를 useEffect 안에서 바로 콜하셔도 괜찮을 것 같습니다 🙂

hansanguk0222 replied:
완료햇어요!

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/ChannelList.tsx:
```diff
@@ -7,28 +7,38 @@ import { useChannel, useAuth } from '@/hooks';
           )
         ) : (
-          <ChannelItem idx={idx} key={channel.id} />
+          ''
```
이부분 정리가 필요해보이네요

삼항 연산자를 너무 쓰면 논리가 잘 안보이고 실수하기 쉽습니다 😭

지금 불필요 하게 길어지는 부분을 아래 코드 참조해서 제거해보세요!

```ts
조건 ? <컴포넌트 /> : ''  // ---> 특정 조건에서 랜더링만 하면되고, 조건이 false일때 랜더링하지 않아도 된다면

조건 && <컴포넌트 /> // ----> 이렇게 하시면 훨씬 간단합니다.
```

hansanguk0222 replied:
고민좀하고 반영하겟습니다!

rnjshippo replied:
반영완료~

rnjshippo commented on client/src/components/ChannelListBox/ChannelListHeader/ChannelListHeader.tsx:
```diff
@@ -1,17 +1,12 @@
   pick: boolean;
-  theme: any;
   ref: any;
 }
```
이 코드 현재 파일에서 사용하나요? 제 눈에는 안보이는데 혹시 안쓴다면 제거해주세요~

hansanguk0222 replied:
이거 수정햇어요!

rnjshippo commented on client/src/components/ChannelListBox/ChannelListHeader/ChannelListHeader.tsx:
```diff
@@ -64,55 +59,70 @@ const PopupItem = styled.div`
+        <SubWrapper>
+          <Button onClick={clickChannel}>{channelListVisible ? '▽' : '▷'}</Button>
+          <Content onClick={clickChannel}>
+            {channelType === 1 ? 'Channels' : 'Direct Messages'}
```
1은 채널을 의미하죠?

utils/constants 폴더로 가실때입니다 ㅎㅎ

거기서 아래 코드처럼 작성하고 export해서 constants를 사용하는 방식으로 바꿔주세요 🙂

그러면 코드가 훨씬 명확해집니다

```ts
const CHANNEL_TYPE = {
   CHANNEL : 1,
   DM : 2 
}
```

hansanguk0222 replied:
상수처리를 생각 못햇네요ㅜ

rnjshippo commented on client/src/components/ChannelListBox/ChannelListHeader/ChannelListHeader.tsx:
```diff
@@ -64,55 +59,70 @@ const PopupItem = styled.div`
+            <Button>+</Button>
+            <Popup pick={addChannelsModalVisible} ref={ref}>
+              <PopupItem onClick={clickAddChannelModal}>
+                {channelType === 1 ? 'Add Channels' : '임시 유저 추가'}
```
이곳도 동일합니다~

hansanguk0222 replied:
수정햇습니다

rnjshippo commented on client/src/components/ChannelModal/AddTopicModal.tsx:
```diff
@@ -85,25 +92,37 @@ const SubmitButton = styled.button`
+  setAddTopicModalVisible,
+}: {
+  setAddTopicModalVisible: Function;
+}): React.FC | any => {
```
any를 쓰신이유가 있으신가요?

이렇게하면 그냥 any를 쓰는것과 다르지 않습니다 😭

잘 모르겠다면 내일 같이 고쳐봐요~

hansanguk0222 replied:
ReactElement로 고쳤습니다.

rnjshippo commented on client/src/components/ChannelModal/AddTopicModal.tsx:
```diff
@@ -1,18 +1,25 @@
+interface Props {
+  ref: any;
+}
+
```
Container에 ref만 지정할때도 에러가 뜨나요?

이 interface를 전달하지 않아도 ref는 동작할 것 같은데.. 

이렇게 하지말고 ref를 선언할 때 아래 처럼 해보세요.
```ts
const ref = useRef<엘리먼트타입>(null);
```

예시) div에 지정할 때
```ts
const containerRef = useRef<HTMLDivElement>(null);
```

hansanguk0222 replied:
Props를 여전히 요구하고 잇어서 얘기를 해봐야 할 꺼 같아요

hansanguk0222 replied:
완료했습니다

rnjshippo commented on client/src/components/ChannelModal/AddTopicModal.tsx:
```diff
@@ -1,18 +1,25 @@
 `;
 
-const Container = styled.div`
+const Container = styled.div<Props>`
```
위에서 언급한것처럼 쓰면 Props를 전달하지 않아도 됩니다~

hansanguk0222 replied:
완료했어요!

rnjshippo commented on client/src/components/ChannelModal/AddTopicModal.tsx:
```diff
@@ -85,25 +92,37 @@ const SubmitButton = styled.button`
   };
-  const onSubmit = (e: React.MouseEvent<HTMLElement>) => {
+
+  useOnClickOutside(ref, () => handler());
```
const handler가 다른곳에서 쓰이지 않는다면 그냥 없애고 이렇게 하시면 될 것 같습니다 🙂

```ts
useOnClickOutside(ref, () => setAddTopicModalVisible(false) );
```

아래 코드까지 읽으니깐 onCancel이랑 handler랑 똑같은 함수네요?

그러면 아래처럼 하시는게 좋겠네요

```ts
useOnClickOutside(ref, onCancel);
```

hansanguk0222 replied:
완료 했어요!

rnjshippo commented on client/src/components/ChannelModal/AddUsersModal.tsx:
```diff
@@ -1,20 +1,26 @@
+
+interface Props {
+  ref: any;
+}
```
위에서 언급한것과 동일합니다. 이 인터페이스는 없애고 useRef를 선언할 때 타입을 지정해주세요~

hansanguk0222 replied:
완료했어요!

rnjshippo commented on client/src/components/ChannelModal/AddUsersModal.tsx:
```diff
@@ -1,20 +1,26 @@
 `;
 
-const Container = styled.div`
+const Container = styled.div<Props>`
```
여기도 지우고~

hansanguk0222 replied:
완료했어요!

rnjshippo commented on client/src/components/ChannelModal/AddUsersModal.tsx:
```diff
@@ -82,3 +103,39 @@ const Remove = styled.button`
+}: {
+  setAddUserModalVisible: Function;
+}): React.FC | any => {
+  const ref = useRef();
```
``` ts
useRef<HTMLDivElement>(null);
```

hansanguk0222 replied:
완료했어요

rnjshippo commented on client/src/components/ChannelModal/CreateChannelModal.tsx:
```diff
@@ -210,15 +216,14 @@ const CreateChannelModal = () => {
-  const createChannel = async (e: React.MouseEvent<HTMLButtonElement>) => {
-    e.preventDefault();
+  const createChannel = async () => {
     const isPublic = secret ? 0 : 1;
```
secret은 boolean 값이죠? 그러면 아래처럼~

```ts
const isPublic = !secret;
```

그냥 아예 isPublic 지우고 secret으로 조건 체크하셔도 상관없긴합니다 ㅎㅎ 보기 편하신대로~

rnjshippo commented on client/src/components/ChannelModal/ShowUsersModal.tsx:
```diff
@@ -91,42 +94,46 @@ const Remove = styled.button`
   const clickAddUser = () => {
-    dispatch(openShowUsers());
-    dispatch(openAddUser());
+    console.log(addUsersModalVisible);
```
나중에 콘솔지워주시고~~

rnjshippo commented on client/src/components/ChannelModal/ShowUsersModal.tsx:
```diff
@@ -91,42 +94,46 @@ const Remove = styled.button`
-      </Container>
-    </ModalBackground>
+    <>
+      {addUsersModalVisible && <AddUsersModal setAddUserModalVisible={setAddUsersModalVisible} />}
```
여긴 &&를 잘썼네요!

다른곳도 리팩하면서 이렇게 바꿔봅시다 👏

rnjshippo commented on client/src/components/LeftSideBar/LeftSideBarContent/LeftSideBarContent.tsx:
```diff
@@ -9,7 +9,8 @@ const Container = styled.div`
     <Container>
-      <ChannelListBox />
+      <ChannelListBox channelType={1} />
+      <ChannelListBox channelType={2} />
```
위에서 언급했듯이 constants로 대체해봅시다!

rnjshippo commented on client/src/components/ThreadListBox/ThreadListHeader/ThreadListHeader.tsx:
```diff
@@ -70,44 +72,59 @@ const RightButton = styled.button`
 `;
 
-const ThreadListHeader = () => {
+const ThreadListHeader = (): React.FC | any => {
```
이것도 any 없애고 적절하게 변환해봐요~

사실 컴포넌트는 typescript가 알아서 추론하도록 두는게 편하긴 합니다

rnjshippo commented on client/src/components/ThreadListBox/ThreadListHeader/ThreadListHeader.tsx:
```diff
@@ -70,44 +72,59 @@ const RightButton = styled.button`
+          <LeftTitle>
+            {current.isPublic} {current.name}
+          </LeftTitle>
+          {current.channelType === 1 && (
```
여기도 constant!

hansanguk0222 replied:
수정햇습니다

rnjshippo commented on client/src/components/ThreadListBox/ThreadListHeader/ThreadListHeader.tsx:
```diff
@@ -70,44 +72,59 @@ const RightButton = styled.button`
+          )}
+        </Left>
+        <Right>
+          {current.channelType === 1 && (
```
요기도~

hansanguk0222 replied:
수정했습니다

rnjshippo commented on client/src/components/common/RightSideBar/RightSideBar.tsx:
```diff
@@ -6,8 +6,8 @@ const Container = styled.div`
-  Header: JSX.Element;
-  Body: JSX.Element;
+  Header: React.FC;
+  Body: React.FC;
```
애매하죠?

react.FC라고 쓰면 제대로 FC가 아닐때 오류가 터집니다 ㅎㅎ

React.ReactNode가 적절하겠네요

https://simsimjae.tistory.com/426

rnjshippo reviewed (commented):
리팩토링하느라 고생많으셨습니다!

내일 나머지 같이 정리하고 또 달려봅시다 👏

hansanguk0222 committed:
[feat] 채널 상수를 추가

hansanguk0222 committed:
[refactor]
1. dispatch를 useEffect안에서 바로 선언하여 실행하기로 함
2. 컴포넌트 반환 타입을 명시

hansanguk0222 committed:
[refactor] 컴포넌트 타입 반환을 명시함

hansanguk0222 committed:
[refactor]
1. 채널 타입을 상수로 받아서 씀
2. 컴포넌트 반환 타입을 명시함

hansanguk0222 committed:
[refactor] 필요없는 부분 삭제

hansanguk0222 committed:
[refactor]
1. 컴포넌트 반환 타입을 명시함
2. 모달 훅(dim을 누르면 꺼지게)을 장착
3. 인자로 주는 Function 타입을 명시
4. ref를 정확히 명시하여서 인터페이스에 ref타입을 넣지 않아도 동작

hansanguk0222 committed:
[refactor] 삼항 연산자가 필요한 곳만 사용
